pkg/webhook/resources/virtualmachine/validator.go

```diff
@@ -75,6 +75,11 @@ func (v *vmValidator) Create(_ *types.Request, newObj runtime.Object) error {
 	if err := v.checkVMSpec(vm); err != nil {
 		return err
 	}
+
```

Comment:
1. CPU and memory quota are checked in below path, what's the consideration to isolate `v.checkStorageResourceQuota` away from it?

```
func (v *vmValidator) checkVMSpec(vm *kubevirtv1.VirtualMachine) error {

	return v.rqCalculator.CheckIfVMCanStartByResourceQuota(vm)
```

2. CPU and memory check are done via `CheckIfVMCanStartByResourceQuota` which skips those stopped VMs; but for storage, it is a bit more complex:  the PVC has kind of `later binding` or `instant binding` strategy.

Are only running VM's PVC are computed or all? or take care of the PVC bind strategy?

Reply:
> CPU and memory quota are checked in below path, what's the consideration to isolate v.checkStorageResourceQuota away from it?

I moved the storage check to a separate function because it has to be implemented differently for VM `Create` and `Update`. I did not want to pass the required parameters through dozens of functions to the corresponding `containsEnoughResourceQuotaToStartVM`. In this case, several functions would have had to be adapted. This way it is nicely separated by logic and no function headers had to be adapted just so that a function 3 call levels lower gets the required parameters.

Reply:
> CPU and memory check are done via CheckIfVMCanStartByResourceQuota which skips those stopped VMs; 

That's right and that's an additional reason why I have also outsourced the functionality in a separate function. All VMs are taken into action in the storage quota assessment. See test case 3 which takes this into account.

Reply:
> it is a bit more complex: the PVC has kind of later binding or instant binding strategy.

The assessment is based on the `harvesterhci.io/volumeClaimTemplates` annotation. It is therefore irrelevant when the storage is created.

If this is to be taken into account, then this must still be implemented; but then I assume that the quota check/assessment cannot be done BEFORE the VM is created.

The question here is how complex the whole thing should be and whether we want it to be.